### PR TITLE
Refs #532:Add passthrough keybindings, make ctrl-c work in terminal

### PR DESCRIPTION
**Pull request by simark**
Currently, typing ctrl-c in the terminal does not yield the expected
result, which is sending a SIGINT to the running process.  The problem
is that a Theia keybinding is registered for ctrl-c (to copy).  Theia
catches it and thinks you want to copy.

When the focus is on a terminal, we don't want Theia to handle the event
itself, but rather let the event propagate.  xterm.js' keyboard event
handler will catch it, and things will work as usual.  To that end,
this patch introduces the concept of passthrough keybindings.  When
the user invokes such a keybinding, Theia does nothing and lets the
event propagate.  To specify a passthrough keybinding, one needs to use
"passthrough" as the command id, which is treated in a special way by
the KeybindingRegistry.

However, we also need the concept of event priority, to make sure that
when we are in a terminal context, that keybinding will take precedence
over the global one.  As a simple approximation, I made the
KeybindingRegistry consider bindings that specify a context before those
that don't.

The terminal module registers passthrough commands for all the key
combination xterm.js recognizes right now:

  https://github.com/xtermjs/xterm.js/blob/v3/src/Terminal.ts#L1684

Some improvements should be made for Mac though.  To match xterm.js'
behavior, the alt-<key> combinations should only be registered on
non-Mac computers, but I don't know how to do that.  Also, I suppose
that on Mac, the ctrl-<key> bindings should be registered with the M4
modifier?  Finally, there's the cmd-A shortcut registered specifically
for Mac.  It would be nice if somebody that has access to a Mac could
improve that.

Signed-off-by: Simon Marchi <simon.marchi@ericsson.com>

### CODE REVIEW

File: packages/core/src/common/keybinding.ts [hunk 1]
```diff
@@ -132,50 +144,83 @@ export class KeybindingRegistry {
                 return;
             }
         }
-        const { keyCode, commandId } = binding;
```
Comment (svenefftinge) [inline packages/core/src/common/keybinding.ts]: why did you remove this?

File: packages/core/src/common/keybinding.ts [hunk 2]
```diff
@@ -132,50 +144,83 @@ export class KeybindingRegistry {
...
-            return bindings[0];
+    getKeybindingsForKeyCode(keyCode: KeyCode, options: GetKeyBindingsOptions = GetKeyBindingsOptionsDefault): Keybinding[] {
+        let bindings = this.keybindings[keyCode.keystroke] || [];
+
+        if (options.activeOnly) {
+            bindings = bindings.filter((k: Keybinding) => this.isActive(k));
         }
-        return bindings.find(this.isActive.bind(this));
+
+        this.sortKeybindingsByPriority(bindings);
```
Comment (svenefftinge) [inline packages/core/src/common/keybinding.ts]: I would consider an overwritten keybinding to be `inactive` so they should be filtered.

File: packages/core/src/common/keybinding.ts [hunk 3]
```diff
@@ -187,20 +232,33 @@ export class KeybindingRegistry {
...
-        if (context && context.isEnabled(binding)) {
-            const handler = this.commandRegistry.getActiveHandler(binding.commandId);
-            if (handler) {
-                event.preventDefault();
-                event.stopPropagation();
-                handler.execute();
+        const bindings = this.getKeybindingsForKeyCode(keyCode);
+
+        for (const binding of bindings) {
+            /* Only execute if it has no context (global contet) or if we're in
```
Comment (svenefftinge) [inline packages/core/src/common/keybinding.ts]: contet -> context

File: packages/terminal/src/browser/terminal-frontend-contribution.ts [hunk 4]
```diff
@@ -54,14 +69,91 @@ export class TerminalFrontendContribution implements CommandContribution, MenuCo
...
+
+             See: https://github.com/xtermjs/xterm.js/blob/v3/src/Terminal.ts#L1684 */
+
+        /* Register ctrl + k (the passed Key) as a passthrough command in the
+           context of the terminal.  */
+        const regCtrl = (k: Key) => {
+            keybindings.registerKeybindings({
+                commandId: KeybindingRegistry.PASSTHROUGH_PSEUDO_COMMAND,
+                keyCode: KeyCode.createKeyCode({ first: k, modifiers: [Modifier.M1] }),
+                context: this.keybindingContextRegistry.getContext(TERMINAL_ACTIVE_CONTEXT),
```
Comment (svenefftinge) [inline packages/terminal/src/browser/terminal-frontend-contribution.ts]: Since it has not been used so far, could we change this property to type string. I.e. you assign the id?

File: packages/terminal/src/browser/terminal-frontend-contribution.ts [hunk 5]
```diff
@@ -30,13 +32,26 @@ export namespace TerminalCommands {
     };
 }
 
+const TERMINAL_ACTIVE_CONTEXT = "terminalActive";
+
+/* Return true if the event targets xterm's internal textarea.  */
+
+function isTerminalFocused(binding: Keybinding, event: KeyboardEvent): boolean {
+    if (event.target instanceof HTMLTextAreaElement) {
```
Comment (akosyakov) [inline packages/terminal/src/browser/terminal-frontend-contribution.ts]: You should be able to check it with `shell.currentWidget instanceof TerminalWidget` instead of exposing `KeyboardEvent`.

File: packages/core/src/common/keybinding.ts (hunk 1 above)
Comment (simark) [reply packages/core/src/common/keybinding.ts]: I just thought that the characters saved by using local variables weren't really worth using the indirection.  But I can add it back if you prefer.

File: packages/core/src/common/keybinding.ts (hunk 2 above)
Comment (simark) [reply packages/core/src/common/keybinding.ts]: Ok, but we should first add documentation to the isActive method to know what it really means for a keybinding to be "active".  Right now, it maps 1:1 to whether the associated command is active.  I don't mind if it changes, but it should be clarified.

File: packages/core/src/common/keybinding.ts (hunk 3 above)
Comment (simark) [reply packages/core/src/common/keybinding.ts]: Oops, thanks.

File: packages/terminal/src/browser/terminal-frontend-contribution.ts (hunk 5 above)
Comment (simark) [reply packages/terminal/src/browser/terminal-frontend-contribution.ts]: How do I get a reference to `shell`?

Comment (simark) [reply packages/terminal/src/browser/terminal-frontend-contribution.ts]: Ahh, through `this.app.shell`.

File: packages/terminal/src/browser/terminal-frontend-contribution.ts (hunk 4 above)
Comment (simark) [reply packages/terminal/src/browser/terminal-frontend-contribution.ts]: Yes, it will be simpler as well.

**Comment by simark** [comment]
I uploaded a new version.  Pending issues (waiting for your feedback) are:

* Using the destructuring to local variables or not.
* Changing the meaning of an "active" keybinding.

File: packages/core/src/common/keybinding.ts (hunk 1 above)
Comment (svenefftinge) [reply packages/core/src/common/keybinding.ts]: No, it's ok. But it is a matter of taste (both ways ok) so I would not have touched it.

File: packages/core/src/common/keybinding.ts (hunk 2 above)
Comment (svenefftinge) [reply packages/core/src/common/keybinding.ts]: Sure, please add documention.

**Comment by svenefftinge** [comment]
I think the issue with `active`here is that we use it for both methods `getKeybindingsForCommand` and `getKeybindingsForKeycode`.
I understand that we want to display the keybinding for a certain currently inactive command in the menu and have the option to get keybindings also for inactive commands. (note here the `active` concept is about the command).

OTOH in we don't know what `active` means for `getKeybindingsForKeycode`. I think it got introduced because it looked symmetric. But it is really wrong, because the `active` means the command in the first and there is no command in the second.

Unless someone comes up with a compelling usecase / reason, we should remove the options from `getKeybindingsForKeycode` and make it again return only one or none keybinding (singular), because there can never be multiple keybindings for the same keycode in any situation.   If there are then we have a conflict and we should handle it somehow. So the implementation should also apply the filtering based on contexts.

**Comment by simark** [comment]
> I think the issue with activehere is that we use it for both methods getKeybindingsForCommand and getKeybindingsForKeycode.
> I understand that we want to display the keybinding for a certain currently inactive command in the menu and therefore have the option to get keybindings also for inactive commands. (note here the active concept is about the command).
> 
> OTOH in we don't know what active means for getKeybindingsForKeycode. I think it got introduced because it looked symmetric. But it is really wrong, because the active means the command in the first and there is no command in the second.
> 
> Unless someone comes up with a compelling usecase / reason, we should remove the options from getKeybindingsForKeycode and make it again return only one or none keybinding (singular), because there can never be multiple keybindings for the same keycode in any situation. If there are then we have a conflict and we should handle it somehow. So the implementation should also apply the filtering based on contexts.

I think it's fine for getKeybindingsForKeycode to return the list of all bindings matching a key stroke, in order of priority, and have the caller decide what to do with the result (e.g. only care about the first one).  If we have many places that do that (select the first one), we could always add another method for convenience that does that.  But for now I think it's fine.

What should happen when you have a keybinding with a context that is enabled (e.g. corresponding view in focus), but bound to an inactive command (e.g. because you don't have an object of the right type selected to do that command)?  Let's say there's a global keybinding (no context) bound to the same key, should we stop execution on the first keybinding and say that because it's bound to an inactive command, the keystroke does nothing?  Or should we carry on to the second keybinding and execute it?

Let's say a particular view has a command that is active when an object of a certain type is selected and inactive otherwise.  If you have that view in focus, but don't have an object of that type selected, I think it would make more sense if it did nothing than if it invoked the global keybinding.  A bit like trying to click on a greyed out menu item.

Or maybe I have the wrong idea of what "active" means for a command?

**Comment by svenefftinge** [comment]
I agree that a registered keybinding for a currently active context should consume the keybinding no matter the associated command is active or not. 
That means we need to check the context before we check the active status, no?

**Comment by simark** [comment]
> I agree that a registered keybinding for a currently active context should consume the keybinding no matter the associated command is active or not.

Ok, thanks for confirming.

> That means we need to check the context before we check the active status, no?

Yes, sounds like it.  Having the activeOnly parameter to getKeybindingsForKeycode leads to the opposite (we filter by active commands first).  In the end I think I'll just remove the activeOnly parameters to the `getKeybindings*` methods entirely, it will be simpler.

**Comment by svenefftinge** [comment]
Note, that we have them on `getKeybindingByCommand` because we want to display a keybinding even for inactive commands in menus. That use case should be supported.

**Comment by simark** [comment]
So in the end there's no use case I can see for `activeOnly == true`.

**Comment by simark** [comment]
I uploaded a new version that:

* removes the activeOnly parameters
* stops looking for more keybindings if a keybinding is in context, but the command is disabled

File: packages/core/src/common/keybinding.ts [hunk 6]
```diff
@@ -137,45 +137,66 @@ export class KeybindingRegistry {
...
-        const bindings = this.commands[commandId];
-        if (!bindings) {
-            return undefined;
-        }
-        if (!options.active) {
-            return bindings[0];
-        }
-        return bindings.find(this.isActive.bind(this));
+    getKeybindingsForCommand(commandId: string): Keybinding[] {
+        return this.commands[commandId] || [];
```
Comment (svenefftinge) [inline packages/core/src/common/keybinding.ts]: Ideally, this should also respect the context. I.e. if you open command palette from within terminal it should not show the disabled keybindings. I think we can do that in a subsequent PR, though.

File: packages/core/src/common/keybinding.ts [hunk 7]
```diff
@@ -187,20 +208,44 @@ export class KeybindingRegistry {
...
+        for (const binding of bindings) {
+            const context = binding.contextId
+                ? this.contextRegistry.getContext(binding.contextId)
+                : undefined;
+
+            /* Only execute if it has no context (global context) or if we're in
+               that context.  */
+            if (!context || context.isEnabled(binding)) {
+
+                if (binding.commandId === KeybindingRegistry.PASSTHROUGH_PSEUDO_COMMAND) {
```
Comment (svenefftinge) [inline packages/core/src/common/keybinding.ts]: use `this.isPseudoCommand`

File: packages/core/src/common/keybinding.ts (hunk 6 above)
Comment (simark) [reply packages/core/src/common/keybinding.ts]: Ok.

File: packages/core/src/common/keybinding.ts (hunk 7 above)
Comment (simark) [reply packages/core/src/common/keybinding.ts]: Done.
